pkg/backup/remap_crd_version_action.go

```diff
@@ -103,24 +110,51 @@ func (a *RemapCRDVersionAction) Execute(item runtime.Unstructured, backup *v1.Ba
 		if c.Type == apiextv1.NonStructuralSchema {
 			log.Debug("CRD is a non-structural schema")
 
-			if err := setV1beta1Version(item); err != nil {
+			item, err = fetchV1beta1CRD(crd.Name, a.betaCRDClient)
+			if err != nil {
 				return nil, nil, err
 			}
-
 			break
 		}
 	}
 
+	// CRD has the field "Spec.PreserveUnknownFields". This meant it was a v1beta1 CRD and cannot be restored as a v1 CRD - validation webhooks will fail.
+	// TODO: add a test case
+	if crd.Spec.PreserveUnknownFields {
+		log.Debug("CRD is set to preserve unknown fields, should be backed up as v1beta1")
+		item, err = fetchV1beta1CRD(crd.Name, a.betaCRDClient)
+		if err != nil {
+			return nil, nil, err
+		}
+
+	}
+
 	return item, nil, nil
 }
 
-// setV1beta1Version updates the apiVersion field of an Unstructured CRD to be the v1beta1 string instead of v1.
-func setV1beta1Version(u runtime.Unstructured) error {
-	// Since we can't manipulate an Unstructured's Object map directly, get a copy to manipulate before setting it back
-	tempMap := u.UnstructuredContent()
-	if err := unstructured.SetNestedField(tempMap, "apiextensions.k8s.io/v1beta1", "apiVersion"); err != nil {
-		return errors.Wrap(err, "unable to set apiversion to v1beta1")
+func fetchV1beta1CRD(name string, betaCRDClient apiextv1beta1client.CustomResourceDefinitionInterface) (*unstructured.Unstructured, error) {
+	betaCRD, err := betaCRDClient.Get(name, metav1.GetOptions{})
+	if err != nil {
+		return nil, errors.Wrapf(err, "error fetching v1beta1 version of %s", name)
+	}
+
+	// Individual items fetched from the API don't always have the kind/API version set
+	// See https://github.com/kubernetes/kubernetes/issues/3030. Unsure why this is happening here and not in main Velero;
+	// probably has to do with List calls and Dynamic client vs typed client
+	if betaCRD.Kind == "" {
```

Comment:
Would `betaCRD.Kind` be anything other than `kuberesource.CustomResourceDefinitions.Resource`
Should we just set these fields instead of conditionally setting them?

Reply:
I don't think it should or could be anything else. Setting the kind and API version shouldn't be an issue, since it should be the same even if it was set by the API server, so I'll remove the conditions around the assignments.